File: iosApp/iosApp/Features/Login/LoginViewModel.swift
import SwiftUI
import Shared

@MainActor @Observable
final class LoginViewModel {
    var email: String = ""
    var password: String = ""
    var loading: Bool = false
    var error: String?

    /// In corso un accesso con passkey (bottone separato, spinner separato).
    var passkeyLoading: Bool = false

    /// Messaggio **neutro**, non un errore: il percorso passkey non deve mai
    /// colorare di rosso la schermata. Quando fallisce, si dice all'utente che
    /// puo' usare la password e si va avanti.
    var passkeyNotice: String?

    private let passkeyAuthenticator = PasskeyAuthenticator()

    init() {
        // Debug seed for visual iteration only.
        if ProcessInfo.processInfo.arguments.contains("--preview-prefill") {
            email = "socio@mensa.it"
            password = "••••••••••"
        }
        // Dev convenience: prefill from env vars MENSA_EMAIL / MENSA_PASSWORD
        // when present (set via SIMCTL_CHILD_* on simulator launch).
        let env = ProcessInfo.processInfo.environment
        if let e = env["MENSA_EMAIL"], !e.isEmpty { email = e }
        if let p = env["MENSA_PASSWORD"], !p.isEmpty { password = p }
    }

    /// Auto-submit if the env asked for it. Called from the view's .task so
    /// we are off the init phase and Task { await ... } is safe.
    func autoLoginIfRequested() async {
        let env = ProcessInfo.processInfo.environment
        Log.auth.info("autoLogin check: MENSA_AUTOLOGIN=\(env["MENSA_AUTOLOGIN"] ?? "<nil>") email=\(self.email.isEmpty ? "empty" : "set") pwd=\(self.password.isEmpty ? "empty" : "set")")
        guard env["MENSA_AUTOLOGIN"] == "1" else { return }
        guard !email.isEmpty, !password.isEmpty else { return }
        Log.auth.info("autoLogin: triggering login")
        _ = await login()
    }

    /// Returns true on success. On failure sets `error`.
    /// AuthRepository.login(email:password:) returns Result<UserModel> in Kotlin.
    /// In Swift/ObjC bridging it throws on failure and returns UserModel on success.
    /// Accesso con passkey. Ritorna true solo in caso di successo.
    ///
    /// Ogni esito negativo riporta l'utente alla password senza errori rossi:
    /// e' l'unico comportamento sensato, perche' chi non ha una passkey su questo
    /// dispositivo non ha sbagliato nulla.
    func loginWithPasskey() async -> Bool {
        let address = email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !address.isEmpty, !passkeyLoading, !loading else { return false }

        passkeyLoading = true
        passkeyNotice = nil
        error = nil
        defer { passkeyLoading = false }

        // Un solo ritentativo, e solo per challenge scaduta: significa che
        // l'utente ha esitato troppo davanti al prompt, non che qualcosa e' rotto.
        for attempt in 0..<2 {
            // `try?`: le suspend fun Kotlin arrivano in Swift come `async throws`
            // e possono lanciare su cancellazione. Il repository non propaga
            // errori applicativi — quelli sono gia' status — quindi un throw qui
            // e' eccezionale e va trattato come fallimento generico.
            guard let start = try? await koin.passkeys.beginLogin(email: address) else {
                passkeyNotice = notice(forStart: PasskeyStartStatus.networkError)
                return false
            }
            guard start.status == PasskeyStartStatus.ready, let challenge = start.challenge else {
                passkeyNotice = notice(forStart: start.status)
                return false
            }

            let credentialJSON: String
            do {
                credentialJSON = try await passkeyAuthenticator.assert(
                    requestOptionsJSON: challenge.requestOptionsJson
                )
            } catch PasskeyAuthenticator.Failure.cancelledOrNoCredential {
                // Annullato dall'utente oppure nessuna passkey locale: iOS non
                // li distingue, e per noi la conseguenza e' la stessa.
                passkeyNotice = tr(
                    "app.login.passkey.use_password",
                    fallback: "Accedi con la password su questo dispositivo."
                )
                return false
            } catch {
                Log.auth.error("Passkey assertion failed: \(error.localizedDescription)")
                passkeyNotice = tr(
                    "app.login.passkey.failed",
                    fallback: "Non riesco a usare la passkey. Prova con la password."
                )
                return false
            }

            guard let result = try? await koin.passkeys.finishLogin(
                loginId: challenge.loginId,
                credentialJson: credentialJSON
            ) else {
                passkeyNotice = notice(forFinish: PasskeyLoginStatus.networkError)
                return false
            }

            if result.status == PasskeyLoginStatus.success {
                Log.auth.info("Passkey login successful")
                return true
            }
            if result.status == PasskeyLoginStatus.challengeExpired, attempt == 0 {
                continue
            }

            passkeyNotice = notice(forFinish: result.status)
            return false
        }

        passkeyNotice = tr(
            "app.login.passkey.failed",
            fallback: "Non riesco a usare la passkey. Prova con la password."
        )
        return false
    }

    private func notice(forStart status: PasskeyStartStatus) -> String {
        if status == PasskeyStartStatus.networkError {
            return tr(
                "app.login.passkey.offline",
                fallback: "Non riesco a contattare il server. Prova con la password."
            )
        }
        return tr(
            "app.login.passkey.none",
            fallback: "Nessuna passkey attiva per questa email. Accedi con la password."
        )
    }

    private func notice(forFinish status: PasskeyLoginStatus) -> String {
        if status == PasskeyLoginStatus.networkError {
            return tr(
                "app.login.passkey.offline",
                fallback: "Non riesco a contattare il server. Prova con la password."
            )
        }
        return tr(
            "app.login.passkey.failed",
            fallback: "Non riesco a usare la passkey. Prova con la password."
        )
    }

    func login() async -> Bool {
        guard !email.isEmpty, !password.isEmpty else { return false }
        loading = true
        error = nil
        passkeyNotice = nil
        defer { loading = false }
        do {
            // Swift bridged name: login(email:password:) — async throws -> UserModel
            _ = try await koin.auth.login(email: email, password: password)
            Log.auth.info("Login successful")
            return true
        } catch {
            Log.auth.error("Login failed: \(error.localizedDescription)")
            self.error = (error as NSError).localizedDescription
            return false
        }
    }
}

File: androidApp/src/main/kotlin/it/mensa/app/features/auth/LoginViewModel.kt
package it.mensa.app.features.auth

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import it.mensa.app.support.Logger
import it.mensa.app.support.koinAccess
import it.mensa.shared.auth.passkey.PasskeyLoginStatus
import it.mensa.shared.auth.passkey.PasskeyStartStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val loading: Boolean = false,
    val error: String? = null,
    /** Accesso con passkey in corso: bottone e spinner separati da quelli password. */
    val passkeyLoading: Boolean = false,
    /**
     * Nota **neutra**, non un errore: il percorso passkey non deve mai colorare
     * di rosso la schermata. Chi non ha una passkey su questo dispositivo non ha
     * sbagliato niente, gli si dice solo di usare la password.
     */
    val passkeyNotice: String? = null,
    /** False sotto API 28: il bottone passkey non va nemmeno mostrato. */
    val passkeySupported: Boolean = PasskeyAuthenticator.isSupported,
)

class LoginViewModel : ViewModel() {

    private val auth = koinAccess().auth
    private val passkeys = koinAccess().passkeys

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    fun onEmailChange(value: String) {
        _uiState.update { it.copy(email = value, error = null, passkeyNotice = null) }
    }

    fun onPasswordChange(value: String) {
        _uiState.update { it.copy(password = value, error = null) }
    }

    fun onLoginClick() {
        val state = _uiState.value
        if (state.email.isBlank() || state.password.isBlank()) return
        if (state.loading || state.passkeyLoading) return

        viewModelScope.launch {
            _uiState.update { it.copy(loading = true, error = null, passkeyNotice = null) }
            val result = runCatching {
                auth.login(state.email.trim(), state.password)
            }
            result.fold(
                onSuccess = { loginResult ->
                    loginResult.fold(
                        onSuccess = {
                            Logger.i("LoginVM", "login", "successful")
                            // Auth state flow in RootViewModel will trigger phase → Main/Onboarding
                        },
                        onFailure = { err ->
                            Logger.e("LoginVM", "login", "failed: ${err.message}")
                            _uiState.update {
                                it.copy(
                                    loading = false,
                                    error = err.message ?: "Errore di accesso",
                                )
                            }
                        },
                    )
                },
                onFailure = { err ->
                    Logger.e("LoginVM", "login", "exception: ${err.message}")
                    _uiState.update {
                        it.copy(
                            loading = false,
                            error = err.message ?: "Errore di rete",
                        )
                    }
                },
            )
        }
    }

    /**
     * Accesso con passkey. Serve solo l'email, perche' la Session API di Zitadel
     * pretende il check utente prima di poter creare la challenge WebAuthn:
     * il login usernameless non e' disponibile.
     *
     * [activityContext] deve essere il Context dell'Activity: Credential Manager
     * ci mostra sopra il prompt di sistema.
     *
     * Ogni esito negativo riporta alla password senza errori rossi.
     */
    fun onPasskeyClick(activityContext: Context) {
        val state = _uiState.value
        val email = state.email.trim()
        if (email.isBlank() || state.loading || state.passkeyLoading) return

        viewModelScope.launch {
            _uiState.update { it.copy(passkeyLoading = true, error = null, passkeyNotice = null) }
            val authenticator = PasskeyAuthenticator(activityContext)

            // Un solo ritentativo, e solo per challenge scaduta: vuol dire che
            // l'utente ha esitato davanti al prompt, non che qualcosa e' rotto.
            repeat(MAX_PASSKEY_ATTEMPTS) { attempt ->
                val start = passkeys.beginLogin(email)
                val challenge = start.challenge
                if (start.status != PasskeyStartStatus.READY || challenge == null) {
                    settle(noticeForStart(start.status))
                    return@launch
                }

                when (val prompt = authenticator.assert(challenge.requestOptionsJson)) {
                    is PasskeyPromptResult.Success -> {
                        val result = passkeys.finishLogin(challenge.loginId, prompt.credentialJson)
                        when (result.status) {
                            PasskeyLoginStatus.SUCCESS -> {
                                Logger.i("LoginVM", "passkey", "successful")
                                // Come per la password: RootViewModel reagisce
                                // al cambio di authState e cambia fase.
                                _uiState.update { it.copy(passkeyLoading = false) }
                                return@launch
                            }

                            PasskeyLoginStatus.CHALLENGE_EXPIRED -> {
                                if (attempt == 0) return@repeat
                                settle(NOTICE_FAILED)
                                return@launch
                            }

                            else -> {
                                Logger.e("LoginVM", "passkey", "finish: ${result.status}")
                                settle(noticeForFinish(result.status))
                                return@launch
                            }
                        }
                    }

                    // L'utente ha chiuso il prompt: nessun messaggio, e' una scelta.
                    PasskeyPromptResult.Cancelled -> {
                        _uiState.update { it.copy(passkeyLoading = false) }
                        return@launch
                    }

                    PasskeyPromptResult.NoCredential -> {
                        settle("Nessuna passkey su questo dispositivo. Accedi con la password.")
                        return@launch
                    }

                    PasskeyPromptResult.Unsupported -> {
                        settle("Questo dispositivo non supporta le passkey. Accedi con la password.")
                        return@launch
                    }

                    // Non dovrebbe capitare in login (excludeCredentials vale solo
                    // in registrazione), ma se capita non va trattato come errore.
                    PasskeyPromptResult.AlreadyRegistered -> {
                        settle(NOTICE_FAILED)
                        return@launch
                    }

                    is PasskeyPromptResult.Failed -> {
                        Logger.e("LoginVM", "passkey", "prompt: ${prompt.message}")
                        settle(NOTICE_FAILED)
                        return@launch
                    }
                }
            }
        }
    }

    fun dismissPasskeyNotice() {
        _uiState.update { it.copy(passkeyNotice = null) }
    }

    private fun settle(notice: String) {
        _uiState.update { it.copy(passkeyLoading = false, passkeyNotice = notice) }
    }

    private fun noticeForStart(status: PasskeyStartStatus): String = when (status) {
        PasskeyStartStatus.NETWORK_ERROR -> NOTICE_OFFLINE
        // UNAVAILABLE copre "email sconosciuta" e "nessuna passkey": il server li
        // rende indistinguibili di proposito, per non dire a chi chiede quali
        // email sono socie.
        else -> "Nessuna passkey attiva per questa email. Accedi con la password."
    }

    private fun noticeForFinish(status: PasskeyLoginStatus): String = when (status) {
        PasskeyLoginStatus.NETWORK_ERROR -> NOTICE_OFFLINE
        else -> NOTICE_FAILED
    }

    private companion object {
        const val MAX_PASSKEY_ATTEMPTS = 2
        const val NOTICE_OFFLINE = "Non riesco a contattare il server. Accedi con la password."
        const val NOTICE_FAILED = "Non riesco a usare la passkey. Accedi con la password."
    }
}
